llvm/include/llvm/Analysis/TargetTransformInfo.h

```diff
@@ -740,6 +743,9 @@ class TargetTransformInfo {
   /// When successful, makes the primary IV dead.
   bool shouldFoldTerminatingConditionAfterLSR() const;
 
+  /// Return true if LSR should run on this loop even if it isn't innermost.
+  bool isProfitableOuterLSR(const Loop &L) const;
```

Comment:
UPSTREAM WHEN YOU CAN: Looks reasonable, ideally you should add a test case with an existing target but that may be difficult as in you may not want to add an implement of `isProfitableOuterLSR` to an existing target.

Reply:
I think this is to make outerloops downcounting, so that we can have our 'secondary' JumpNonZeroDecrement construction generated for them.  Since this needs to have a pointer live across a two-deep loopnest to be profitable, gains are marginal and regressions likely.